docs/implementation_design/object_storage.md

```diff
@@ -1 +1,87 @@
+For the persistent storage in a database it uses package `badger` from hive.go. It is a simple and fast key-value database that performs well for both reads and writes simultaneously.
+
+Both solutions are implemented in `database` package, along prefixes definitions that can be used during creation of new object storage elements.
+
+Database plugin is responsible for creating `store` instance of the chosen database under the directory specified with `CfgDatabaseDir` parameter. It will manage proper closure of the database upon recieving shutdown signal. During the start configuration, database is marked as unhealthy, it will be marked as healthy on shutdown, then garbage collector is run and database can be closed.
+
+## ObjectStorage
+
+
+Assuming we need to store data for some newly created object `A`. We need to define a new prefix for our package in the `database` package, and prefixes for single storage objects. They will be later used during objectstorage cration. A package prefix will be combined with store specific prefix to create specyfic realm.
+```Go
+package example
+
+type Storage struct {
+	A 						*objectstorage.ObjectStorage
+	...
+	shutdownOnce			sync.Once
+}
+```
+### ObjectStorage factory
+To easily create multiple storage objects instances for one package, the most convinient way is to use facory function.
+```Go
+osFactory := objectstorage.NewFactory(store, database.Prefix)
+```
+It needs two parameters:
+- `store` - the key value `kvstore` object
+- `database.Prefix` - prefix defined in the `database` package for our new `example` package. It will be responsible for automatic configuration of newly provided KVStore object.
+
+
+After defining storage factory for the group of object we can use it to create `*objectstorage.ObjectStorage` instance:
+```Go
+AStorage = osFactory.New(objPrefix, FromObjectStorage)
+AStorage = osFactory.New(objPrefix, FromObjectStorage, optionalOptions...)
+```
+For the function parameter we should provide:
+- `objPrefix` - mentioned before, we provide the object specyfic prefix.
+- `FromObjectStorage` function, a `StorableObject` factory that receives a key and the serialized data of the object and returns an empty `StorableObject` with just key set.
+- `optionalOptions` -  an optional parameter provided in a form of options array `[]objectstorage.Option`. All possible options are defined in `objectstorage.Options`. If we do not specify them during creation, the default values will be used, such as enabling persistance or setting cache time to 0.
+
+### StorableObject
+`StorableObject` is an interface that allow the dynamic creation of different object types depending on the stored data. We need to make sure that all methods required by the interface are implemented to use object storage factory.
```

Comment:
that allows the dynamic
to use (a or the?) object storage factory

Reply:
corrected